torch/distributed/_sharded_tensor/api.py

```diff
@@ -551,6 +558,176 @@ def sharding_spec(self) -> ShardingSpec:
+            indices: List[List[int]] = [[0]] * world_size
+            # When we do the chunk split, we always ensure the first N - 1 chunks get max out
+            # and then the Nth chunk gets the rest. So input_split_sizes like [3, 3, 3, 4]
+            # are not possible. The expected split size will be [4, 4, 4, 1].
+            sharded_dim_size_max = max(input_split_sizes)
+            for idx, placement in enumerate(resharding_spec.placements):
+                split_size = input_split_sizes[placement.rank()]  # type: ignore[union-attr, index]
+                offset_start_idx = idx * sharded_dim_size_max
+                indices[placement.rank()] = list(  # type: ignore[union-attr, index]
+                    range(offset_start_idx, offset_start_idx + split_size)
+                )
+            indices_flatten = list(idx for indice in indices for idx in indice)
+            local_shard = local_shard.index_select(
+                reshard_dim, torch.tensor(indices_flatten, device=local_shard.device)
+            )
+
+        # Because reshard_dim != original shard_dim. We need to compute the
+        # size of tensor from each rank.
+        output_tensor_list = [torch.tensor(1)] * world_size
+        split_size = get_split_size(self.size(current_sharding_dim), world_size)
+        rearrange_output_list = False
+        indices = []
+        for idx, placement in enumerate(self._sharding_spec.placements):
+            sharded_dim_size = get_chunked_dim_size(
+                self.size(current_sharding_dim), split_size, idx
+            )
+            output_tensor_size = list(local_shard.size())
+            output_tensor_size[current_sharding_dim] = sharded_dim_size
+            output_tensor_size[reshard_dim] = input_split_sizes[current_rank]
+            output_tensor_list[placement.rank()] = (  # type: ignore[union-attr, index]
+                torch.empty(output_tensor_size, device=local_shard.device)
+            )
+            indices.append(placement.rank())  # type: ignore[union-attr, index, arg-type]
+            if idx != placement.rank():  # type: ignore[union-attr]
+                rearrange_output_list = True
+
+        # Perform autograd enabled all2all.
+        input_tensor_list = torch.split(local_shard, input_split_sizes, dim=reshard_dim)
+        input_tensor_list = [tensor.contiguous() for tensor in input_tensor_list]
+        output_tensor_list = all_to_all(
+            output_tensor_list,
+            input_tensor_list,
+            group=self._process_group,
+        )
+
+        if rearrange_output_list:
+            # Need to re-arrange original shard_dim of output_tensor_list.
+            output_tensor_list = [output_tensor_list[idx] for idx in indices]  # type: ignore[call-overload]
+        local_shard = torch.cat(output_tensor_list, dim=current_sharding_dim)
+        self.local_shards()[0].tensor = local_shard
+        return self
+
+    def local_shard(self) -> torch.Tensor:
```

Comment:
Or maybe name it `local_shard_tensor`? We can avoid multiple calls like `local_shards()[0]`. cc: @pritamdamania87

Reply:
yeah it's a common operation, `local_shard_tensor` lgtm.

Reply:
The purpose was to avoid a lot of duplicate code we have today where we validate there is a single shard and then do `local_shards()[0].tensor`. I think instead of `local_shard_tensor`, maybe we just call it `local_tensor`.

Reply:
I see. OK, let's use a shorter name then.